File: nexago_app/lib/features/arenas/data/promotions_repository.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import '../domain/arena_promotion.dart';

class PromotionsRepository {
  PromotionsRepository(this._firestore);

  final FirebaseFirestore _firestore;

  Stream<List<ArenaPromotion>> watchPromotions(String arenaId) {
    final id = arenaId.trim();
    if (id.isEmpty) return Stream.value(const []);

    return _firestore
        .collection('arenas')
        .doc(id)
        .collection('promotions')
        .where('active', isEqualTo: true)
        .snapshots()
        .map((snap) {
      final list = snap.docs.map(ArenaPromotion.fromFirestore).toList();
      list.sort((a, b) => a.label.compareTo(b.label));
      return list;
    });
  }

  /// Todas as promoções (gestor), inclusive inativas.
  Stream<List<ArenaPromotion>> watchAllPromotions(String arenaId) {
    final id = arenaId.trim();
    if (id.isEmpty) return Stream.value(const []);

    return _firestore
        .collection('arenas')
        .doc(id)
        .collection('promotions')
        .snapshots()
        .map((snap) {
      final list = snap.docs.map(ArenaPromotion.fromFirestore).toList();
      _sortPromotionsForManager(list);
      return list;
    });
  }

  Future<void> setPromotionActive({
    required String arenaId,
    required String promotionId,
    required bool active,
  }) async {
    final arena = arenaId.trim();
    final promo = promotionId.trim();
    if (arena.isEmpty || promo.isEmpty) {
      throw ArgumentError('arenaId e promotionId são obrigatórios.');
    }

    await _firestore
        .collection('arenas')
        .doc(arena)
        .collection('promotions')
        .doc(promo)
        .update({'active': active});
  }

  Future<String> createPromotion({
    required String arenaId,
    required ArenaPromotion promotion,
  }) async {
    final ref = await _firestore
        .collection('arenas')
        .doc(arenaId.trim())
        .collection('promotions')
        .add(promotion.toFirestore());
    return ref.id;
  }

  Future<void> updatePromotion({
    required String arenaId,
    required String promotionId,
    required ArenaPromotion promotion,
  }) async {
    final arena = arenaId.trim();
    final promo = promotionId.trim();
    if (arena.isEmpty || promo.isEmpty) {
      throw ArgumentError('arenaId e promotionId são obrigatórios.');
    }

    await _firestore
        .collection('arenas')
        .doc(arena)
        .collection('promotions')
        .doc(promo)
        .update(promotion.toFirestoreUpdate());
  }

  Future<void> deletePromotion({
    required String arenaId,
    required String promotionId,
  }) async {
    final arena = arenaId.trim();
    final promo = promotionId.trim();
    if (arena.isEmpty || promo.isEmpty) {
      throw ArgumentError('arenaId e promotionId são obrigatórios.');
    }

    await _firestore
        .collection('arenas')
        .doc(arena)
        .collection('promotions')
        .doc(promo)
        .delete();
  }

  static void _sortPromotionsForManager(List<ArenaPromotion> list) {
    list.sort((a, b) {
      if (a.active != b.active) return a.active ? -1 : 1;
      return a.label.compareTo(b.label);
    });
  }
}

File: frontend/projects/arena/src/app/painel/promotions/promotions-repository.ts
import { arenaPromotionFromFirestore, type ArenaPromotion } from '@nexago/arena-discovery';
import { addDoc, collection, deleteDoc, deleteField, doc, getDocs, serverTimestamp, Timestamp, updateDoc, type Firestore } from 'firebase/firestore';

/** Espelha `PromotionsRepository` (Flutter) — `arenas/{arenaId}/promotions/{promoId}`. Criar/editar
 *  exige plano Pro/Parceiro (`arenaEntitled`, ver `ArenaContextService.hasCapability('promocoes')`);
 *  excluir é livre. O desconto é XOR: só um entre `discountPercent`/`fixedPricePerHourReais`. */

export interface PromotionInput {
  label: string;
  active: boolean;
  /** Vazio = todas as quadras da arena. */
  courtIds: string[];
  /** ISO 1-7 (seg-dom). Vazio = todos os dias. */
  weekdays: number[];
  startTime: string;
  endTime: string;
  discountPercent: number | null;
  fixedPricePerHourReais: number | null;
  validFrom: Date | null;
  validUntil: Date | null;
}

function validatePromotionInput(input: PromotionInput): string | null {
  if (!input.label.trim()) return 'Informe o nome da promoção.';
  if (input.discountPercent == null && input.fixedPricePerHourReais == null) {
    return 'Informe um desconto (percentual ou valor fixo por hora).';
  }
  if (!input.startTime || !input.endTime) return 'Informe o horário de início e fim.';
  return null;
}

function basePayload(input: PromotionInput): Record<string, unknown> {
  return {
    active: input.active,
    label: input.label.trim(),
    courtIds: input.courtIds,
    weekdays: input.weekdays,
    startTime: input.startTime,
    endTime: input.endTime,
  };
}

export async function fetchAllPromotions(db: Firestore, arenaId: string): Promise<ArenaPromotion[]> {
  const snap = await getDocs(collection(db, 'arenas', arenaId, 'promotions'));
  const list = snap.docs.map((d) => arenaPromotionFromFirestore(d));
  list.sort((a, b) => {
    if (a.active !== b.active) return a.active ? -1 : 1;
    return a.label.localeCompare(b.label, 'pt-BR', { sensitivity: 'base' });
  });
  return list;
}

export async function createPromotion(db: Firestore, arenaId: string, input: PromotionInput): Promise<string> {
  const error = validatePromotionInput(input);
  if (error) throw new Error(error);

  const payload: Record<string, unknown> = {
    ...basePayload(input),
    createdAt: serverTimestamp(),
    updatedAt: serverTimestamp(),
  };
  if (input.discountPercent != null) payload['discountPercent'] = input.discountPercent;
  if (input.fixedPricePerHourReais != null) payload['fixedPricePerHourReais'] = input.fixedPricePerHourReais;
  if (input.validFrom) payload['validFrom'] = Timestamp.fromDate(input.validFrom);
  if (input.validUntil) payload['validUntil'] = Timestamp.fromDate(input.validUntil);

  const ref = await addDoc(collection(db, 'arenas', arenaId, 'promotions'), payload);
  return ref.id;
}

export async function updatePromotion(db: Firestore, arenaId: string, promoId: string, input: PromotionInput): Promise<void> {
  const error = validatePromotionInput(input);
  if (error) throw new Error(error);

  await updateDoc(doc(db, 'arenas', arenaId, 'promotions', promoId), {
    ...basePayload(input),
    updatedAt: serverTimestamp(),
    discountPercent: input.discountPercent != null ? input.discountPercent : deleteField(),
    fixedPricePerHourReais: input.fixedPricePerHourReais != null ? input.fixedPricePerHourReais : deleteField(),
    validFrom: input.validFrom ? Timestamp.fromDate(input.validFrom) : deleteField(),
    validUntil: input.validUntil ? Timestamp.fromDate(input.validUntil) : deleteField(),
  });
}

export async function setPromotionActive(db: Firestore, arenaId: string, promoId: string, active: boolean): Promise<void> {
  await updateDoc(doc(db, 'arenas', arenaId, 'promotions', promoId), { active, updatedAt: serverTimestamp() });
}

export async function deletePromotion(db: Firestore, arenaId: string, promoId: string): Promise<void> {
  await deleteDoc(doc(db, 'arenas', arenaId, 'promotions', promoId));
}
